internal/api/v2/handler_payments_create.go

```diff
@@ -0,0 +1,226 @@
+	if r.Reference == "" {
+		return errors.New("reference is required")
+	}
+
+	if r.ConnectorID == "" {
+		return errors.New("connectorID is required")
+	}
+
+	if r.CreatedAt.IsZero() || r.CreatedAt.After(time.Now()) {
+		return errors.New("createdAt is empty or in the future")
+	}
+
+	if r.Amount == nil {
+		return errors.New("amount is required")
+	}
+
+	if r.Type == "" {
+		return errors.New("type is required")
+	}
+
+	if _, err := models.PaymentTypeFromString(r.Type); err != nil {
+		return fmt.Errorf("invalid type: %w", err)
+	}
+
+	if r.Scheme == "" {
+		return errors.New("scheme is required")
+	}
+
+	if _, err := models.PaymentSchemeFromString(r.Scheme); err != nil {
+		return fmt.Errorf("invalid scheme: %w", err)
+	}
+
+	if r.Asset == "" {
+		return errors.New("asset is required")
+	}
+
+	if r.Status == "" {
+		return errors.New("status is required")
+	}
+
+	if _, err := models.PaymentStatusFromString(r.Status); err != nil {
+		return fmt.Errorf("invalid status: %w", err)
+	}
+
+	if r.SourceAccountID != nil {
+		_, err := models.AccountIDFromString(*r.SourceAccountID)
+		if err != nil {
+			return fmt.Errorf("invalid sourceAccountID: %w", err)
+		}
+	}
+
+	if r.DestinationAccountID != nil {
+		_, err := models.AccountIDFromString(*r.DestinationAccountID)
+		if err != nil {
+			return fmt.Errorf("invalid destinationAccountID: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func paymentsCreate(backend backend.Backend) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx, span := otel.Tracer().Start(r.Context(), "v2_paymentsCreate")
+		defer span.End()
+
+		var req createPaymentRequest
+		err := json.NewDecoder(r.Body).Decode(&req)
+		if err != nil {
+			otel.RecordError(span, err)
+			api.BadRequest(w, ErrMissingOrInvalidBody, err)
+			return
+		}
+
+		if err := req.validate(); err != nil {
+			otel.RecordError(span, err)
+			api.BadRequest(w, ErrValidation, err)
+			return
+		}
+
+		connectorID := models.MustConnectorIDFromString(req.ConnectorID)
+		paymentType := models.MustPaymentTypeFromString(req.Type)
+		status := models.MustPaymentStatusFromString(req.Status)
+		raw, err := json.Marshal(req)
+		if err != nil {
+			otel.RecordError(span, err)
+			api.InternalServerError(w, r, err)
+			return
+		}
+		pid := models.PaymentID{
+			PaymentReference: models.PaymentReference{
+				Reference: req.Reference,
+				Type:      paymentType,
+			},
+			ConnectorID: connectorID,
+		}
+
+		payment := models.Payment{
+			ID:            pid,
+			ConnectorID:   connectorID,
+			Reference:     req.Reference,
+			CreatedAt:     req.CreatedAt.UTC(),
+			Type:          paymentType,
+			InitialAmount: req.Amount,
+			Amount:        req.Amount,
+			Asset:         req.Asset,
+			Scheme:        models.MustPaymentSchemeFromString(req.Scheme),
+			SourceAccountID: func() *models.AccountID {
+				if req.SourceAccountID == nil {
+					return nil
+				}
+				return pointer.For(models.MustAccountIDFromString(*req.SourceAccountID))
```

Comment:
This function panics so it's really not appropriate within a http handler. Especially since it's evaluating a string that was provided via user input.

We need to parse this earlier in the validation step and return a 400 if the account IDs are invalid.

Reply:
It's already parsed inside the validate function

Reply:
Then you should be using the parsed value from the validation rather than parsing again, no?